### PR TITLE
Add ADR for sidecar upgrade job

### PR DESCRIPTION
**Pull request by XanderStrike**
[#174283800](https://www.pivotaltracker.com/story/show/174283800)

### Acceptance Steps

1. Read the doc
2. Understand

cc @ndhanushkodi @tcdowney

### CODE REVIEW

File: doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+container image that contains the Job's logic.
+
+All istio components will be tagged with their Istio version so that the job can
+positively determine that the correct version of control plane components are
+alive and healthy. We will also name the job according to it's Istio version, so
+that we can take advantage of `Jobs` inherent immutability in cases where a
+cf-for-k8s upgrade does not contain a new Istio version (pushing the same job
+again will not cause it to rerun, preventing workloads from rolling
+unnecessarily). Subsequent jobs will clean up previous ones.
+
```
Comment (ndhanushkodi) [inline doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md:46]: Maybe we could add a bit about how the Job gets triggered, and how it waits on the labeled istio resources to reach a new version before then restarting the workloads. One thing that I keep having to remind myself about is that we wait for 2 configmaps before restarting the workloads, and would be nice to capture that here.

**Review by ndhanushkodi — CHANGES_REQUESTED** [review]

File: doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md (hunk 1 above)
Comment (XanderStrike) [reply doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md:46]: We don't do that any more! See https://github.com/cloudfoundry/cf-k8s-networking/commit/cbade83a3c344918c1b580e1ebedb0a6d986c345

I wonder how nitty gritty we want to be in these ADRs. I worry about too closely coupling the ADR to details of the implementation that may change over time and the ADR becoming stale. This is a good example of that, and we may later find a better way of achieving our goal of waiting util control plane components are ready. I wouldn't want to have to come back and change the ADR in that case, so I tried to phrase it in more of an outcome centered way than an implementation way.

File: doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+## Decision
+
+We will use the kubernetes
+[Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/) resource
+to run the kubectl command needed to roll workload pods, after waiting for the
+new Istio control plane to be up and healthy.
+
+To that end, we will add the necessary minimal `ServiceAccounts` and `Roles`
+needed to list resources in the `istio-system` namespace, and restart resources
```
Comment (tcdowney) [inline doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md:35]: The mechanism described here should serve our needs for now, so I'm not suggesting any changes here.

But there are other reasons to restart the app pods as well (such as whatever the equivalent of a rootfs rollout is these days). I wonder if one day in the future we can all unify on a mechanism to trigger these.

Somethin' to chew on!!!! 🍽

cc @cwlbraa

File: doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md (hunk 1 above)
Comment (ndhanushkodi) [reply doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md:46]: Ah oops! And that totally makes sense. Resolving

**Review by ndhanushkodi — APPROVED** [review]

**Review by tcdowney — APPROVED** [review]

File: doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md (hunk 2 above)
Comment (cwlbraa) [reply doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md:35]: Today in capiland, we're modifying the OPI-templated StatefulSet spec to trigger controlled rootfs rollouts, but StatefulSets know nothing about istio injection, right?

**Pull request merged**